tools/wptrunner/wptrunner/executors/executorwktr.py

```diff
@@ -0,0 +1,282 @@
+        while True:
+            line = _read_line(self.stdout_queue, deadline, "latin-1", logger=self.logger).rstrip()
+
+            if line == "Content-Type: text/plain":
+                return self._read_text_block(deadline)
+
+            if line == "Content-Type: image/png":
+                return self._read_image_block(deadline)
+
+            if line == "#EOF":
+                return None
+
+    def _read_text_block(self, deadline=None):
+        """Tries to read a plain text block in utf-8 encoding before the `deadline`.
+        """
+        result = ""
+
+        while True:
+            line = _read_line(self.stdout_queue, deadline, "utf-8", "replace", False, logger=self.logger)
+
+            if line.endswith(self.eof_marker):
+                result += line[:-len(self.eof_marker)]
+                break
+            elif line.endswith('#EOF\r\n'):
+                result += line[:-len('#EOF\r\n')]
+                self.logger.warning('Got a CRLF-terminated #EOF - this is a driver bug.')
+                break
+
+            result += line
+
+        return result
+
+    def _read_image_block(self, deadline=None):
+        """Tries to read an image block (as a binary png) before the `deadline`.
+        """
+        content_length_line = _read_line(self.stdout_queue, deadline, "utf-8", logger=self.logger)
+        assert content_length_line.startswith("Content-Length:")
+        content_length = int(content_length_line[15:])
+
+        result = bytearray()
+
+        while True:
+            line = _read_line(self.stdout_queue, deadline, raise_crash=False, logger=self.logger)
+            excess = len(line) + len(result) - content_length
+
+            if excess > 0:
+                # This is the line that contains the EOF marker.
+                assert excess == len(self.eof_marker)
+                result += line[:-excess]
+                break
+
+            result += line
+
+        return result
+
+
+class WKTRErrorsPart(ProtocolPart):
+    """This protocol part is responsible for collecting the errors reported by wktr.
+    """
+    name = "wktr_errors"
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.stderr_queue = parent.browser.stderr_queue
+
+    def read_errors(self):
+        """Reads the entire content of the stderr queue as is available right now (no blocking).
+        """
+        result = ""
+
+        while not self.stderr_queue.empty():
+            # There is no potential for race conditions here because this is the only place
+            # where we read from the stderr queue.
+            result += _read_line(self.stderr_queue, None, "utf-8", "replace", False, logger=self.logger)
+
+        return result
+
+
+class WKTRProtocol(Protocol):
+    implements = [WKTRTestPart, WKTRErrorsPart]
+    init_timeout = 10  # Timeout (seconds) to wait for #READY message.
+
+    def connect(self):
+        """Waits for wktr to emit its "#READY" message which signals that it is fully
```

Comment:
It is surprising to me that this function has a docstring claiming it does something, but no body.

Reply:
Yeah, I should probably go and fix all the comments and docstrings here :|